### PR TITLE
Change/use persisted pointer for fast header

### PR DESCRIPTION
**Pull request by asdacap**
- Currently unlike oldbodies, orldreceipt or beacon header, the pointer for fast header is initialized via binary search from genesis to sync pivot. 
- This cause a problem with era/portal which may insert header in between arbitrarily.
- This change the pointer to be persisted like beacon header and the pointer is changed manually in fast header instead of on every block/header insert.
- Also don't download already persisted header. This only 2x faster than downloading though.

#### What types of changes does your code introduce?

- [X] Optimization
- [X] Refactoring

#### Requires testing

- [X] Yes
- [ ] No

#### If yes, did you write tests?

- [X] Yes
- [ ] No

#### Notes on testing

- Mainnet sync work.

### CODE REVIEW

File: src/Nethermind/Nethermind.Synchronization/FastBlocks/FastBlocksBatch.cs [hunk 1]
```diff
@@ -81,7 +81,9 @@ public double? WaitingTime
             => (_handlingStartTime ?? _stopwatch.ElapsedMilliseconds) - (_waitingStartTime ?? _handlingStartTime ?? _stopwatch.ElapsedMilliseconds);
         public double? HandlingTime
             => (_handlingEndTime ?? _stopwatch.ElapsedMilliseconds) - (_handlingStartTime ?? _stopwatch.ElapsedMilliseconds);
-        public long? MinNumber { get; set; }
+
+        /// Minimum header number for allocation strategy.
+        public virtual long? MinNumber { get; }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/FastBlocks/FastBlocksBatch.cs:86]: should it be abstract?

File: src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs [hunk 2]
```diff
@@ -334,13 +339,20 @@ private bool HasDependencyToProcess
...
+            {
+                batch = new();
+                batch.StartNumber = Math.Max(HeadersDestinationNumber,
+                    _lowestRequestedHeaderNumber - _headersRequestSize);
+                batch.RequestSize = (int)Math.Min(_lowestRequestedHeaderNumber - HeadersDestinationNumber,
+                    _headersRequestSize);
+                _lowestRequestedHeaderNumber = batch.StartNumber;
+
+                batch = ProcessPersistedPortion(batch);
+            } while (batch is null && ShouldBuildANewBatch());
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs:355]: This is sligthly confusing, shouldn't the loop be outside of this method?

File: src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs [hunk 3]
```diff
@@ -456,14 +466,69 @@ private static HeadersSyncBatch BuildDependentBatch(HeadersSyncBatch batch, long
...
+
+            using ArrayPoolList<BlockHeader> headers = new ArrayPoolList<BlockHeader>(1);
+            headers.Add(lastHeader);
+            for (long i = batch.EndNumber - 1; i >= batch.StartNumber; i--)
+            {
+                // Don't worry about fork, `InsertHeaders` will check for fork and retry if it is not on the right fork.
+                BlockHeader nextHeader = _blockTree.FindHeader(lastHeader.ParentHash!, i);
+                if (nextHeader is null) break;
+                lastHeader = nextHeader;
+                headers.Add(lastHeader);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs:508]: more readable:
```suggestion
                headers.Add(nextHeader);
                lastHeader = nextHeader;
```

File: src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs [hunk 4]
```diff
@@ -456,14 +466,69 @@ private static HeadersSyncBatch BuildDependentBatch(HeadersSyncBatch batch, long
...
+            }
+
+            ArrayPoolList<BlockHeader> reversedBatch = new ArrayPoolList<BlockHeader>(headers.Count);
+            foreach (BlockHeader blockHeader in headers.Reverse())
+            {
+                reversedBatch.Add(blockHeader);
+            }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs:515]: Why need a second `ArrayPoolList`? Add `Reverse` to the class and reverse in-place (same as `List<>`)

File: src/Nethermind/Nethermind.Merge.Plugin/Synchronization/BeaconHeadersSyncFeed.cs [hunk 5]
```diff
@@ -36,7 +36,15 @@ public sealed class BeaconHeadersSyncFeed : HeadersSyncFeed
...
 
-    protected override BlockHeader? LowestInsertedBlockHeader => _blockTree.LowestInsertedBeaconHeader;
+    protected override BlockHeader? LowestInsertedBlockHeader
+    {
+        get => _blockTree.LowestInsertedBeaconHeader;
+        set
+        {
+            // Set in blocktree
+        }
+    }
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Synchronization/BeaconHeadersSyncFeed.cs:46]: Looks fishy, can you elaborate in comment?

File: src/Nethermind/Nethermind.Synchronization/FastBlocks/FastBlocksBatch.cs (hunk 1 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/FastBlocks/FastBlocksBatch.cs:86]: Probably

File: src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs (hunk 2 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs:355]: Moved outside

File: src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs (hunk 4 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Synchronization/FastBlocks/FastHeadersSyncFeed.cs:515]: Changed to `AsSpan().Reverse()`.

**Review by LukaszRozmej — APPROVED** [review]

**Pull request merged**